app/graphql/mutations/update_project_publish_targets.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+class Mutations::UpdateProjectPublishTargets < Mutations::BaseMutation
+  argument :project, Inputs::StringKey, "String key for project", required: true
+  argument(
+    :publish_targets,
+    [Inputs::StringKey],
+    "An array representing the new set of enabled publish targets for the given project, "\
+    "or an empty array to remove all project enabled publish targets.",
+    required: true
+  )
+
+  field :project_publish_targets, [Types::PublishTargetType], null: false
+
+  def resolve(project:, publish_targets:)
+    # This should be an all or nothing update
+    ActiveRecord::Base.transaction do
+      project = Project.find_by!(project.to_h)
+      # Ensure that the user initiating this update is allowed to do so for the given project
+      ability.authorize! :update, project
+
+      publish_targets = PublishTarget.where(string_key: publish_targets.map { |pt| pt['stringKey'] })
+      project.publish_targets = publish_targets.to_a
+      project.save!
+    end
```

Comment:
Does this actually need to be in a transaction, since you're only performing one update query?

Reply:
If for some reason the update does need to be in a transaction, I think you can at least move some of the earlier lines out of it to minimize transaction time.

Reply:
Oh, maybe you're wrapping in a transaction because `project.publish_targets = publish_targets.to_a` and then `project.save!` updates the updated_at timestamp?  In that case, maybe only those two lines need to be in the transaction.

Reply:
Oh, maybe you're wrapping in a transaction because `project.publish_targets = publish_targets.to_a` instantly performs a query and then `project.save!` updates the updated_at timestamp in a separate query?  In that case, I think only those two lines need to be in the transaction.

Reply:
No, nothing like that - I'm adhering to what I think we probably want to encourage as a general pattern in the UI-driven update layer. You're right that we could shave lines off here - as you pointed out to me, the HABTM relationship automatically saves, so you could technically drop that `save!` line, too. But the resulting method would, in my view, reward high-context maintainers (able to infer the HABTM behavior, etc) and punish low-context maintainers ("why aren't they saving this record? how are these tests passing?"), and the latter (ie, team members who do not spend lots of contiguous time in this codebase) is I think the more likely audience. This isn't an absolute position, but it's something to consider: What balance of performance, legibility and maintainability do we want to arrive at? What should we fix with comments/documentation? The balance is much different for batch processing, I think, than in the gql layer of individual updates.

Reply:
I understand what you mean, and it is easy to forget that has_many/habtm relationships are automatically added.  How about we just take the non-updating lines out of the transaction then?  That seems like a good practice regardless of rails-specific knowledge.

Reply:
Sounds like a good compromise!

Reply:
Moved reads/authorize outside of transaction.